torchtext/data/metrics.py

```diff
@@ -0,0 +1,93 @@
+import math
+import collections
+from torchtext.data.utils import ngrams_iterator
+
+
+def bleu_score(candidate_corpus, references_corpus, max_n=4, weights=[0.25] * 4):
+    """Computes the BLEU score between a candidate translation corpus and a references
+    translation corpus. Based on https://www.aclweb.org/anthology/P02-1040.pdf
+
+    Arguments:
+        candidate_corpus: a list of candidate translations. Each translation is a list of
+            tokens
+        references_corpus: a list of lists of reference translations. Each translation
+            is a list of tokens
+        max_n: the maximum n-gram we want to use. E.g. if max_n=3, we will use unigrams,
+            bigrams and trigrams
+        weights: weights used for each n-gram category (uniform by default)
+
+    Examples:
+        >>> candidate_corpus = [['I', 'ate', 'the', 'apple'], ['I', 'did']]
+        >>> references_corpus = [[['I', 'ate', 'it'], ['I', 'ate', 'apples']],
+                [['I', 'did']]]
+        >>> bleu_score(candidate_corpus, references_corpus)
+        >>> 0.7598356856515925
+    """
+
+    def _compute_ngram_counter(tokens, max_n):
+        """ Create a Counter with a count of unique n-grams in the tokens list
+
+        Arguments:
+            tokens: a list of tokens (typically a string split on whitespaces)
+            max_n: the maximum order of n-gram wanted
+
+        Outputs:
+            output: a collections.Counter object with the unique n-grams and their
+                associated count
+
+        Examples:
+            >>> from torchtext.data.functional import _compute_ngram_counter
+            >>> tokens = ['me', 'me', 'you']
+            >>> _compute_ngram_counter(tokens, 2)
+                Counter({('me',): 2,
+                 ('you',): 1,
+                 ('me', 'me'): 1,
+                 ('me', 'you'): 1,
+                 ('me', 'me', 'you'): 1})
+        """
+        assert max_n > 0
+
+        ngrams = [tuple(x.split(' ')) for x in ngrams_iterator(tokens, max_n)]
+        ngrams_counter = collections.Counter(ngrams)
```

Comment:
Counter accepts iterators, so you can use ```collections.Counter(tuple(x.split(' ')) for x in ngrams_iterator(tokens, max_n))```. The advantage is that you don't need to materialize the entire string, so you'll save on memory roundtrips.

Reply:
I think this is a good idea. In general, we only materialize the string if really necessary. Otherwise, we just send them to next generator pipeline.